File: open_vms/public/js/production_plan_bom_link.js

frappe.ui.form.on('Production Plan', {
    get_items(frm) {
        if (frm.doc.get_items_from !== 'Sales Order') return;

        if (!frm.doc.sales_orders || frm.doc.sales_orders.length === 0) {
            frappe.msgprint(__("Please select Sales Orders first using 'Get Sales Orders'"));
            return;
        }

        const sales_orders = frm.doc.sales_orders.map(row => row.sales_order);

        frappe.call({
            method: 'open_vms.public.py.production_plan_bom_link.get_items_with_custom_bom',
            args: {
                production_plan: frm.doc.name,
                sales_orders: sales_orders
            },
            callback: function (r) {
                if (!r.message) {
                    frappe.msgprint(__('No BOMs matched the selected Sales Orders.'));
                    return;
                }

                const bom_map = {};

                // Group BOMs by Sales Order + Item Code
                r.message.forEach(item => {
                    const key = `${item.sales_order}__${item.item_code}`;
                    if (!bom_map[key]) {
                        bom_map[key] = [];
                    }
                    if (item.bom_no) {
                        bom_map[key].push(item.bom_no);
                    }
                });

                // Wait until PO Items are loaded
                setTimeout(() => {
                    const po_items = [...frm.doc.po_items]; // Copy to avoid direct mutation

                    po_items.forEach(row => {
                        const key = `${row.sales_order}__${row.item_code}`;
                        const boms = bom_map[key] || [];

                        if (boms.length > 0) {
                            // Set first BOM on current row
                            frappe.model.set_value(row.doctype, row.name, 'bom_no', boms[0]);

                            // Add additional rows for other BOMs
                            for (let i = 1; i < boms.length; i++) {
                                const new_row = frappe.model.add_child(frm.doc, 'Production Plan Item', 'po_items');
                                new_row.item_code = row.item_code;
                                new_row.item_name = row.item_name;
                                new_row.sales_order = row.sales_order;
                                new_row.bom_no = boms[i];
                                new_row.planned_qty = row.planned_qty;
                                new_row.stock_uom = row.stock_uom;
                                new_row.warehouse = row.warehouse;
                                new_row.planned_start_date = row.planned_start_date;
                                // Add any other fields you want copied here
                            }
                        } else {
                            // Clear BOM if none found
                            frappe.model.set_value(row.doctype, row.name, 'bom_no', '');
                        }
                    });

                    frm.refresh_field('po_items');
                }, 800);
            }
        });
    }
});

File: open_vms/public/py/production_plan_bom_link.py

import frappe
import json
from frappe.utils import now_datetime

no_cache = True

@frappe.whitelist()
def get_items_with_custom_bom(production_plan, sales_orders):
    """
    Return list of BOMs matched per Sales Order and Item.
    For multiple BOMs, return each as a separate row.
    If no BOM found, still return a row with bom_no = "".
    """
    if isinstance(sales_orders, str):
        sales_orders = json.loads(sales_orders)

    matched_boms = []

    for so in sales_orders:
        # Get items from Sales Order
        so_items = frappe.get_all(
            "Sales Order Item",
            filters={"parent": so},
            fields=["*"]
        )
        print(so_items, "&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&")
        for item in so_items:
            bom_list = []

            # Step 1: Get all active BOMs for Sales Order + Item
            boms = frappe.get_all(
                "BOM",
                filters={
                    "custom_sales_order": so,
                    "item": item.item_code,
                    "is_active": 1,
                    "docstatus": 1
                },
                fields=["name"],
                order_by="creation desc"
            )

            if boms:
                bom_list = [b["name"] for b in boms]
            else:
                # Step 2: Get cancelled BOM (original)
                original_bom = frappe.db.get_value(
                    "BOM",
                    {
                        "custom_sales_order": so,
                        "item": item.item_code,
                        "docstatus": 2
                    },
                    "name"
                )

                if original_bom:
                    # Step 3: Get all active amended BOMs from that
                    amended_boms = frappe.get_all(
                        "BOM",
                        filters={
                            "amended_from": original_bom,
                            "sales_order": so,
                            "item": item.item_code,
                            "is_active": 1,
                            "docstatus": 1
                        },
                        fields=["name"],
                        order_by="creation desc"
                    )

                    if amended_boms:
                        bom_list = [b["name"] for b in amended_boms]

            # Append one row per BOM
            if bom_list:
                for bom in bom_list:
                    matched_boms.append({
                        "sales_order": so,
                        "item_code": item.item_code,
                        "item_name": item.item_name,
                        "bom_no": bom,
                        "warehouse": item.get("warehouse"),
                        "planned_qty": item.get("qty"),
                        "stock_uom": item.get("uom"),
                        "planned_start_date": now_datetime()
                    })
            else:
                # Even if no BOM found, still return one row with empty bom_no
                matched_boms.append({
                    "sales_order": so,
                    "item_code": item.item_code,
                    "item_name": item.item_name,
                    "bom_no": "",
                    "warehouse": item.get("warehouse"),
                    "planned_qty": item.get("qty"),
                    "stock_uom": item.get("uom"),
                    "planned_start_date": now_datetime()
                })

    print(matched_boms, "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
    return matched_boms


# import frappe
# no_cache = True


# @frappe.whitelist()
# def get_items_with_custom_bom(production_plan, sales_orders):
#     """Return BOMs filtered by custom_sales_order field"""
#     if isinstance(sales_orders, str):
#         import json
#         sales_orders = json.loads(sales_orders)

#     matched_boms = []

#     for so in sales_orders:
#         # Find BOMs where custom_sales_order matches current SO
#         boms = frappe.get_all(
#             "BOM",
#             filters={
#                 "custom_sales_order": so,
#                 "is_active": 1,
#                 "docstatus": 1
#             },
#             fields=["name", "item", "item_name", "custom_sales_order"]
#         )

#         for bom in boms:
#             matched_boms.append({
#                 "sales_order": so,
#                 "bom_no": bom["name"],
#                 "item_code": bom["item"],
#                 "item_name": bom["item_name"]
#             })
        

#     return matched_boms


